scipy/sparse/tests/test_64bit.py

```diff
@@ -0,0 +1,431 @@
+            m = mat_cls(np.random.rand(11, 1))
+            assert self._compare_index_dtype(m, np.int64)
+
+        for mat_cls in self.MAT_CLASSES:
+            check(mat_cls)
+
+    def test_decorator_maxval_random(self):
+        # Test that the with_64bit_maxval_limit decorator works (2)
+
+        @with_64bit_maxval_limit(random=True)
+        def check(mat_cls):
+            seen_32 = False
+            seen_64 = False
+            for k in range(100):
+                m = mat_cls(np.random.rand(9, 9))
+                seen_32 = seen_32 or self._compare_index_dtype(m, np.int32)
+                seen_64 = seen_64 or self._compare_index_dtype(m, np.int64)
+                if seen_32 and seen_64:
+                    break
+            else:
+                raise AssertionError("both 32 and 64 bit indices not seen")
+
+        for mat_cls in self.MAT_CLASSES:
+            check(mat_cls)
+
+    def test_downcast_intp(self):
+        # Check that bincount and ufunc.reduceat intp downcasts are
+        # dealt with. The point here is to trigger points in the code
+        # that can fail on 32-bit systems when using 64-bit indices,
+        # due to use of functions that only work with intp-size indices.
+
+        @with_64bit_maxval_limit(fixed_dtype=np.int64, downcast_maxval=1)
+        def check_limited(csc_container, csr_container, coo_container):
+            # These involve indices larger than `downcast_maxval`
+            a = csc_container([[1, 2], [3, 4], [5, 6]])
+            pytest.raises(AssertionError, a.count_nonzero, axis=1)
+            pytest.raises(AssertionError, a.sum, axis=0)
+
+            a = csr_container([[1, 2, 3], [3, 4, 6]])
+            pytest.raises(AssertionError, a.count_nonzero, axis=0)
+            pytest.raises(AssertionError, a.sum, axis=1)
+
+            a = coo_container([[1, 2, 3], [3, 4, 5]])
+            pytest.raises(AssertionError, a.count_nonzero, axis=0)
+            a.has_canonical_format = False
+            pytest.raises(AssertionError, a.sum_duplicates)
+
+        @with_64bit_maxval_limit(fixed_dtype=np.int64)
+        def check_unlimited(csc_container, csr_container, coo_container):
+            # These involve indices smaller than `downcast_maxval`
+            a = csc_container([[1, 2], [3, 4], [5, 6]])
+            a.count_nonzero(axis=1)
+            a.sum(axis=0)
+
+            a = csr_container([[1, 2, 3], [3, 4, 6]])
+            a.count_nonzero(axis=0)
+            a.sum(axis=1)
+
+            a = coo_container([[1, 2, 3], [3, 4, 5]])
+            a.count_nonzero(axis=0)
+            a.has_canonical_format = False
+            a.sum_duplicates()
+
+        check_limited(csc_array, csr_array, coo_array)
+        check_unlimited(csc_array, csr_array, coo_array)
+        check_limited(csc_matrix, csr_matrix, coo_matrix)
+        check_unlimited(csc_matrix, csr_matrix, coo_matrix)
+
+
+# We could track includes rather than ignores. But this way no new tests get skipped
+IGNORE_TESTS = {
```

Comment:
Let's leave this out for now, to avoid the risk of accidentally missing coverage in case these methods do start calling `get_index_dtype` later. That's what we currently run anyway, so it's not additional cost vs the current main branch.